### PR TITLE
OCR3 execute plugin: report phase

### PR DESCRIPTION
**Pull request by winder**
Select commit reports from the outcome to include in the final execution report and build the report.

### CODE REVIEW

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 1]
```diff
@@ -8,36 +8,56 @@ import (
...
+	"github.com/smartcontractkit/chainlink-common/pkg/merklemulti"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
 	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
 	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
-
-	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
 )
 
+// maxReportSize that should be returned as an execution report payload.
+const maxReportSize = 250_000
```
Comment (rstout) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: What unit? e.g. bytes? Number of reports? Number of messages?

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 2]
```diff
@@ -190,6 +210,175 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
...
 }
 
+// TokenDataReader is an interface for reading extra token data from an async process.
+// TODO: Build a token data reading process.
+type TokenDataReader interface {
+	ReadTokenData(ctx context.Context, srcChain cciptypes.ChainSelector, num cciptypes.SeqNum) ([][]byte, error)
+}
+
+// buildSingleChainReport constructs a single chain report from a commit report.
+func buildSingleChainReport(ctx context.Context, lggr logger.Logger, hasher cciptypes.MessageHasher, tokenDataReader TokenDataReader, codec cciptypes.ExecutePluginCodec, report cciptypes.ExecutePluginCommitDataWithMessages, maxReportSize, maxMessages int) (cciptypes.ExecutePluginReportSingleChain, int, error) {
```
Comment (rstout) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: Could you break this up into multiple lines?

Comment (rstout) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: maxReportSize doesn't have a type?

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 3]
```diff
@@ -8,36 +8,56 @@ import (
...
 
+// maxReportSize that should be returned as an execution report payload.
+const maxReportSize = 250_000
+
 // Plugin implements the main ocr3 plugin logic.
 type Plugin struct {
-	reportingCfg ocr3types.ReportingPluginConfig
-	cfg          cciptypes.ExecutePluginConfig
-	ccipReader   cciptypes.CCIPReader
+	ctx             context.Context
```
Comment (dimkouv) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: Please double check this.
For example a client might do:

```
ctx = context.WithTimeout(5sec)
plugin, err = newPlugin(ctx)
return plugin
// context is cancelled
```

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 4]
```diff
@@ -8,36 +8,56 @@ import (
...
+	reportingCfg    ocr3types.ReportingPluginConfig
+	cfg             cciptypes.ExecutePluginConfig
+	ccipReader      cciptypes.CCIPReader
+	reportCodec     cciptypes.ExecutePluginCodec
+	msgHasher       cciptypes.MessageHasher
+	tokenDataReader TokenDataReader
```
Comment (dimkouv) [inline core/services/ocr3/plugins/ccip/execute/plugin.go:30]: TokenDataReader could be defined in cl-common similar to the other interfaces.

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 5]
```diff
@@ -8,36 +8,56 @@ import (
...
 	cfg cciptypes.ExecutePluginConfig,
 	ccipReader cciptypes.CCIPReader,
+	reportCodec cciptypes.ExecutePluginCodec,
+	msgHasher cciptypes.MessageHasher,
+	lggr logger.Logger,
 ) *Plugin {
 	lastReportTS := &atomic.Int64{}
 	lastReportTS.Store(time.Now().Add(-cfg.MessageVisibilityInterval).UnixMilli())
 
+	// TODO: initialize tokenDataReader.
```
Comment (dimkouv) [inline core/services/ocr3/plugins/ccip/execute/plugin.go:48]: I think better to pass it as an argument and let caller initialize it.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 2 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: @winder maybe update linter rules to have max 120 chars in a line

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 6]
```diff
@@ -190,6 +210,175 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
...
+	ReadTokenData(ctx context.Context, srcChain cciptypes.ChainSelector, num cciptypes.SeqNum) ([][]byte, error)
+}
+
+// buildSingleChainReport constructs a single chain report from a commit report.
+func buildSingleChainReport(ctx context.Context, lggr logger.Logger, hasher cciptypes.MessageHasher, tokenDataReader TokenDataReader, codec cciptypes.ExecutePluginCodec, report cciptypes.ExecutePluginCommitDataWithMessages, maxReportSize, maxMessages int) (cciptypes.ExecutePluginReportSingleChain, int, error) {
+	if maxMessages == 0 {
+		maxMessages = len(report.Messages)
+	}
+
+	numMsg := int(report.SequenceNumberRange.End() - report.SequenceNumberRange.Start() + 1)
```
Comment (dimkouv) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: Possible to have an overflow, maybe add a defensive check? or use uint64 for `numMsg` ?

Comment (dimkouv) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: nit: `numMsgs` ?

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 7]
```diff
@@ -190,6 +210,175 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
...
+	if numMsg != len(report.Messages) {
+		return cciptypes.ExecutePluginReportSingleChain{}, 0, fmt.Errorf("malformed report %s, unexpected number of messages: expected %d, got %d", report.MerkleRoot.String(), numMsg, len(report.Messages))
+	}
+
+	treeLeaves := make([][32]byte, 0)
```
Comment (dimkouv) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: maybe use `cciptypes.Bytes32` ?

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 8]
```diff
@@ -190,6 +210,175 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
...
+		}
+		leaf, err := hasher.Hash(ctx, msg)
+		if err != nil {
+			return cciptypes.ExecutePluginReportSingleChain{}, 0, fmt.Errorf("unable to hash message (%d, %d): %w", msg.SourceChain, msg.SeqNum, err)
+		}
+		treeLeaves = append(treeLeaves, leaf)
+	}
+
+	lggr.Debugw("constructing merkle tree", "sourceChain", report.SourceChain, "treeLeaves", len(treeLeaves))
+	tree, err := merklemulti.NewTree(hashutil.NewKeccak(), treeLeaves)
```
Comment (dimkouv) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: We'll need to place this in provider interface. Some chain might not support keccak256, same for commit. As a followup

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 2 above)
Comment (dimkouv) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: This method is somewhat large ~90lines. Could be split in separate ones.

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 9]
```diff
@@ -190,6 +210,175 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
...
+		}
+	}
+
+	lggr.Infow("selected messages from commit report for execution", "sourceChain", report.SourceChain, "commitRoot", report.MerkleRoot.String(), "numMessages", len(toExecute), "totalMessages", numMsg, "toExecuteu", len(toExecute))
+	proof, err := tree.Prove(toExecute)
+	if err != nil {
+		return cciptypes.ExecutePluginReportSingleChain{}, 0, fmt.Errorf("unable to prove messages for report %s: %w", report.MerkleRoot.String(), err)
+	}
+
+	var proofsCast []cciptypes.Bytes32
```
Comment (dimkouv) [inline core/services/ocr3/plugins/ccip/execute/plugin.go:367]: this prob not needed if you work with `Bytes32` everywhere

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 7 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: I think I'd have to re-implement the Keccak hasher, it currently uses `[32]byte` which carries over to all the code interacting with proofs. I'll plan to do that which should clean up some of the types in this PR.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 9 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go:367]: I believe it's from `hashutil.NewKeccak()`. [See response above.](https://github.com/smartcontractkit/ccip/pull/1091#discussion_r1654780150)

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 1 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: Good question. Copied this value from the ocr2 implementation. I've assumed bytes in this PR but was planning to double check. I'll clarify the units.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 2 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: This is a go thing, the types can be omitted from consecutive parameters with the same type In this case `maxReportSize` and `maxMessages` are both `int`s. One of go's love-it-or-hate-it features.

[Here's another example](https://github.com/smartcontractkit/ccip/pull/1091/files#diff-ee23d57795b49ea6f214b06581e765a67632febc3ace0ba69f3b43f4d5a19037R193).

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 3 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: Sure. I really hope it isn't used like that. Cancelling this context would be a very convenient way to indicate that the system is shutting down. :sweat_smile:

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 4 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go:30]: I was on the fence about that point. It certainly fits the model, but I believe the current OCR2 plugin puts the CCTP api url in the job spec, doesn't that mean it's self-contained to the plugin?

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 5 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go:48]: [Discussion here.](https://github.com/smartcontractkit/ccip/pull/1091/files#r1654825057)

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 4 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go:30]: Thanks for bringing this up, I [created an issue to follow up on this directly](https://smartcontract-it.atlassian.net/browse/CCIP-2641). Let's focus on the reporting parts on this PR, I'll create issues for any other points like this.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 6 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: I don't think we have to worry about having math.MaxInt messages in one commit report.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 7 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: After reviewing this, I'm going to suggest an alternative:

Replace `Bytes32` with `hashutil.Hash`.

```
// Hash contains all supported hash formats.
// Add additional hash types e.g. [20]byte as needed here.
type Hash interface {
	[32]byte
}
```

seems to be incompatible with 
```
type Bytes32 [32]byte
```

Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: [Created an issue to track this followup task.](https://smartcontract-it.atlassian.net/browse/CCIP-2643)

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 10]
```diff
@@ -199,6 +220,234 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
...
+
+	lggr.Debugw(
+		"constructing merkle tree",
+		"sourceChain", report.SourceChain,
+		"treeLeaves", len(treeLeaves))
+	tree, err := merklemulti.NewTree(hashutil.NewKeccak(), treeLeaves)
+	if err != nil {
+		return cciptypes.ExecutePluginReportSingleChain{}, 0, fmt.Errorf(
+			"unable to constructing merkle tree from messages: %w", err)
+	}
```
Comment (rstout) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: I think this logic should be abstracted into a `buildMerkleTree()` function

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 11]
```diff
@@ -199,6 +220,234 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
...
+		}
+	}
+
+	lggr.Infow(
+		"selected messages from commit report for execution",
+		"sourceChain", report.SourceChain,
+		"commitRoot", report.MerkleRoot.String(),
+		"numMessages", len(toExecute),
+		"totalMessages", numMsgs,
+		"toExecute", len(toExecute))
```
Comment (rstout) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: "numMessages" and "toExecute" log the same value, should we remove one?

Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: Yup, looks like I mixed up numMessages and totalMessages at some point

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 12]
```diff
@@ -199,6 +220,233 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
 	return ocr3types.QuorumFPlusOne, nil
 }
 
+// TokenDataReader is an interface for reading extra token data from an async process.
+// TODO: Build a token data reading process.
+type TokenDataReader interface {
+	ReadTokenData(ctx context.Context, srcChain cciptypes.ChainSelector, num cciptypes.SeqNum) ([][]byte, error)
+}
+
+// buildSingleChainReport constructs a single chain report from a commit report.
```
Comment (rstout) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: I'd find it helpful if this doc line was expanded, e.g. what this function is doing, what a "single chain report" is, etc

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 3 above)
Comment (makramkd) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: Where would this context come from? 

The plugin is shut down by libocr, so having this as part of the plugin doesn't make much sense to me at the moment. You should use contexts provided to each plugin method.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 7 above)
Comment (makramkd) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: Yeah merkle roots may not always be 32 bytes long

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 13]
```diff
@@ -199,6 +220,233 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
...
+	partialReport := false
+	var finalReports []cciptypes.ExecutePluginReportSingleChain
+	for reportIdx, report := range reports {
+		finalReport, encodedSize, err :=
+			buildSingleChainReport(ctx, lggr, hasher, tokenDataReader, codec, report, 0)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to build a single chain report (max): %w", err)
+		}
+
+		// The full report is too large, binary search for best report size
```
Comment (makramkd) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: Inline binary search doesn't seem great maintenance-wise, any way you could use sort.Search instead?

Comment (makramkd) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: Also bsearch is kinda infamous for its edge cases

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 14]
```diff
@@ -244,15 +492,36 @@ func (p *Plugin) Outcome(
 		}
 	}
 
-	// TODO: select reports and messages for the final exec report.
-	// TODO: may only need the proofs for the final exec report rather than the report and the messages.
+	outcomeReports, commitReports, err :=
+		selectReport(p.ctx, p.lggr, p.msgHasher, p.reportCodec, p.tokenDataReader, commitReports, maxReportSizeBytes)
```
Comment (makramkd) [inline core/services/ocr3/plugins/ccip/execute/plugin.go]: Is it easier to build the reports in Outcome rather than report?

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 15]
```diff
@@ -244,15 +492,36 @@ func (p *Plugin) Outcome(
...
+	if err != nil {
+		return nil, err
+	}
+
+	report := []ocr3types.ReportWithInfo[[]byte]{{
+		Report: encoded,
+		Info:   nil,
+	}}
+
+	return report, nil
```
Comment (makramkd) [inline core/services/ocr3/plugins/ccip/execute/plugin.go:538]: I think you produce potentially many reports above, why only choose one and return it from here? We have the ability to return many reports in OCR3, may as well use it right?

Comment (makramkd) [reply core/services/ocr3/plugins/ccip/execute/plugin.go:538]: Or do you plan on revisiting this in the future to take advantage of this?

File: core/services/ocr3/plugins/ccip/execute/plugin.go [hunk 16]
```diff
@@ -199,6 +220,233 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
...
+	if err != nil {
+		lggr.Errorw("unable to encode report", "err", err, "report", finalReport)
+		return cciptypes.ExecutePluginReportSingleChain{}, 0, fmt.Errorf("unable to encode report: %w", err)
+	}
+
+	return finalReport, len(encoded), nil
+}
+
+// selectReport takes an ordered list of reports and selects the first reports that fit within the maxReportSizeBytes.
+func selectReport(
```
Comment (rstout) [inline core/services/ocr3/plugins/ccip/execute/plugin.go:403]: This function took me a long time (~45mins) to understand at an adequate level. I was initially confused by the binary search, initially I thought it was searching for a report in `reports` that could fit in `finalReports`, but it's actually doing binary search within a report to determine how much of the report should be included. I was also tripped up by `reports = reports[fullyExecuted:]` on line 437, it took me a while to convince myself that partially executed reports weren't being dropped, because we break (and stop adding to `finalReports`) when `partialReport = true`. I still don't feel confident that this function is bug-free tbh. I think it'd be helpful to add helper functions, e.g. `func get_partial_report(report, maxSize)` to abstract away the binary search.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 3 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: Looks like we mistakenly included a context object in the `NewPlugin` function and they should be removed. I went this direction because it seemed like this should be some sort of lifecycle context.

There's a problem with using the provided contexts, for some reason `Encode` and `MessageHasher.Hash` both expect a context but `Outcome` doesn't provide one. I suppose this makes sense, since it should be pure. Any idea why `Encode` and `MessageHasher.Hash` would need the context?

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 7 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: @makramkd I guess thats another reason we have to update hashutil. In the same way that the old ford cars could be any color you like so long as it is black, hashutil can use any checksum [so long as its [32]byte](https://github.com/smartcontractkit/chainlink-common/blob/main/pkg/hashutil/hasher.go#L5).

Go's `crypto.Hash` type is defined as a `uint` to accommodate anything from 16-64 bytes, which seems like a fine approach.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 14 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: For Outcome vs Report, it works either way and probably makes a little more sense to put it in the Report phase (on the other hand, it's nice to form complete consensus on the final report).

The problem is if you need to know what was executed during the next observation. If it's not in the outcome then it has to re-run the report selection algorithm.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 15 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go:538]: The "execution report" is literally just an array of those single chain reports. Don't worry, this is everything!
```
	execReport := cciptypes.ExecutePluginReport{
		ChainReports: outcomeReports,
	}
```

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 16 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go:403]: Totally agree with everything you said here.

Considering the complexity of this code, I think the test code is much more readable. I'd encourage you to take a look at some of those test cases, I made an effort to cover all the corner cases I could come up with and have 100% coverage aside from a few obscure error cases (i.e. the crypto libs returning an error).

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 7 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: Created a ticket to cover the digest size work: https://smartcontract-it.atlassian.net/browse/CCIP-2644

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 3 above)
Comment (makramkd) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: I believe those have contexts because they're under the hood gRPC calls to a server that does the encoding and hashing. That said if the encoding and hashing needs to be done in Outcome and Report I don't think having them behind a gRPC call is very performant.

Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: @makramkd we'll have to figure out an alternative (to my implementation or how these functions are implemented). Outcome and Reports are supposed to be pure, presumably that's why they don't provide a context.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 13 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: sort.Search isn't perfect either because it doesn't handle errors. I'll at least break out some helper functions for now.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 2 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: I split this up a little. I don't want to go further because the remaining code needs some refactoring to handle commit reports which the plugin isn't ready to execute yet.

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 3 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: Contexts removed. [New issue and left some TODOs in the code.](https://smartcontract-it.atlassian.net/browse/CCIP-2662)

File: core/services/ocr3/plugins/ccip/execute/plugin.go (hunk 8 above)
Comment (winder) [reply core/services/ocr3/plugins/ccip/execute/plugin.go]: I created some related followups:
* https://smartcontract-it.atlassian.net/browse/CCIP-2643
* https://smartcontract-it.atlassian.net/browse/CCIP-2644

We need to figure out some improvements so that MessageHasher, Bytes32, merklemulti and hashutil all work together.

**Comment by makramkd** [comment]
Reminder to move this PR into https://github.com/smartcontractkit/chainlink-ccip

**Comment by winder** [comment]
Moved to chainlink-ccip
